Ensure `npm bin` is in $PATH when installing globally

ariporad opened the pull request:
Fixes #6993.

Commit:
bin: add bin.ensureInPath

It will call the callback with true if the global npm bin is in $PATH, or false otherwise.

Commit:
install: warn on global install if npm bin is not in $PATH

Fixes #6993

ariporad commented:
Also, the travis build is failing, and I'm 99.99% sure it's not what I did. Could someone look at it for me? Thanks!

iarna commented:
@ariporad It's not, https://github.com/npm/npm/pull/10192 is the fix for it.

iarna commented:
The two test files in https://github.com/ariporad/npm/commit/67c1f193e805a66602447da694b09983d76ab4ae shouldn't be included. Also it looks like you had $PATH expand when you ran commit there?

ariporad commented:
@iarna: Ok, fixed $PATH, removed `/test/tap/dedupe`. Should be ready to go.

One other thing: I realized that with this change, running `npm bin -g` won't warn if the global npm bin isn't in $PATH. I don't know if this is a problem, but if it is, I'll fix it.

othiym23 commented:
The code is looking good, and now we need to add a tests before we merge and release this. Do you want to take a pass at it, and if so, how can we help you get started writing an npm test? (They're a little tricky.)

ariporad commented:
@othiym23: Sure, I'll try it. Where would I start?

ariporad commented:
@iarna, @othiym23: bump: I'd love to try to write some tests for it, where would I start?

iarna commented on lib/bin.js:
```diff
@@ -11,12 +11,21 @@ function bin (args, silent, cb) {
-  if (npm.config.get('global') && PATH.indexOf(b) === -1) {
-    npm.config.get('logstream').write('(not in PATH env variable)\n')
-  }
+bin.ensureInPath = function ensureInPath (cb) {
```
So the first thing I'd do is write a functional test for this part. To do that you'll want to be able to control all of the variables that go into this:
```
var requireInject = require('require-inject')
var binPath = '/usr/local/bin/npm' // or whatever
var envPATH = '/usr/bin:/bin:/usr/local/bin'
var isGlobal = true
var bin = requireInject('../../lib/bin.js', {
  npm: {
    bin: binPath,
    config: {
      get: function (key) {
          if (key !== 'global') throw new Error('unexpected config key: ' + key)
          return isGlobal
      }
    }
  },
  osenv: {
    path: function () {
      return envPATH
    }
  }
})

iarna commented on lib/install.js:
```diff
@@ -289,7 +290,12 @@ Installer.prototype.run = function (cb) {
         log.verbose('error', postInstallEr.stack)
       }
-      cb(installEr || postInstallEr, self.getInstalledModules(), self.idealTree)
+      ensureInPath(function isInPath (isInPathEr, isInPath) {
```
I think this test should probably go in `/lib/install/validate-tree` but I'll grant that we're not exactly validating the tree here, this is more about validating your install environment, so I don't hate having it in `install.js` either. But regardless, it shouldn't be calling `log.warn`, instead it should be constructing an error object and pushing it on to `idealTree.warnings`, which will be printed out after the installation summary.

iarna commented:
Hey @ariporad I finally got a chance to give some hints on writing a test for this. I also had one comment on your fix here.

Tests go in `test/tap/name-of-test.js` – you can look at any of the others there to get a sense of the boiler plate, but it boils down to:
```
var test = require('tap').test
var common = require('../common-npm.js')

test('name-of-group-of-tests', function (t) {
  t.is(actual, expected, 'description of what this is testing')
  t.done()
})
```

iarna commented:
If you've got further questions please feel free to ask here, (or if you want a faster turn around, [DM me on Twitter](https://twitter.com/ReBeccaOrg))

iarna commented:
We're closing this pull request as it has gone sixty days without activity.  In our experience pull requests that go this long are unlikely to land.  If you have updates to it, please feel free to open a new pull request.

For more information about our new issue aging policies and why we've instituted them please see our [blog post](http://blog.npmjs.org/post/161832149430/npm-the-npm-github-issue-tracker-and-you).
